### PR TITLE
Improve connection handling

### PR DESCRIPTION
**Pull request by sigurdm**
Most of this is appropriated from the grpc implementation we had in https://github.com/dart-lang/appengine. Big thanks to @mkustermann for exlpaining the intricate details!

This should make the connection handling a lot more robust, and also faster.
Hopefully dealing with: https://github.com/grpc/grpc-dart/issues/228

### CODE REVIEW

File: lib/src/client/http2_connection.dart [hunk 1]
```diff
@@ -69,25 +73,36 @@ class Http2ClientConnection implements connection.ClientConnection {
...
-    }
     final securityContext = credentials.securityContext;
-    if (securityContext != null) {
-      socket = await SecureSocket.secure(socket,
-          host: authority,
+    Socket socket;
+    if (securityContext == null) {
+      socket = await Socket.connect(host, port);
+    } else {
+      final socket = await SecureSocket.connect(host, port,
```
Comment (tvolkert) [inline lib/src/client/http2_connection.dart]: You're creating a new local variable here, so the outer `socket` remains null.  Perhaps name this `secureSocket`, and then set `socket = secureSocket` after the "h2" if check below?

Comment (tvolkert) [reply lib/src/client/http2_connection.dart]: (also, is there any way to test this code path?)

File: lib/src/client/http2_connection.dart [hunk 2]
```diff
@@ -69,25 +73,36 @@ class Http2ClientConnection implements connection.ClientConnection {
...
       }
     }
-    socket.done.then((_) => _handleSocketClosed());
-    return ClientTransportConnection.viaSocket(socket);
+
+    final connection = ClientTransportConnection.viaSocket(socket);
+    socket.done.then((_) => _abandonConnection());
+
+    // Give the settings settings-frame a bit of time to arrive.
+    await new Future.delayed(_estimatedRoundTripTime);
```
Comment (tvolkert) [inline lib/src/client/http2_connection.dart]: This seems like a smell - can we not be notified somehow when we're ready?

File: lib/src/client/http2_connection.dart [hunk 3]
```diff
@@ -96,17 +111,45 @@ class Http2ClientConnection implements connection.ClientConnection {
...
+  /// Abandons the current connection if it is unhealthy or has been open for
+  /// too long.
+  ///
+  /// Assumes [_transportConnection] is not `null`.
+  void _refreshConnectionIfUnhealthy() {
+    final bool isHealthy = _transportConnection.isOpen;
+    final bool shouldRefresh =
+        _connectionLifeTimer.elapsed > options.connectionTimeout;
+    if (!isHealthy || shouldRefresh) {
+      if (isHealthy) {
```
Comment (tvolkert) [inline lib/src/client/http2_connection.dart]: This is the same as `if (shouldRefresh)`, and there's no reason for it to be in the outer `if` block.

**Comment by tvolkert** [comment]
The old `appengine` code used a Dialer to ensure only one new http/2 connection is made if multiple event loops try to establish a new connection.  Does this handle that case?

File: lib/src/client/http2_connection.dart (hunk 1 above)
Comment (sigurdm) [reply lib/src/client/http2_connection.dart]: Oops
Thanks for catching this!

File: lib/src/client/http2_connection.dart (hunk 2 above)
Comment (jonasfj) [reply lib/src/client/http2_connection.dart]: Maybe add a comment that explains why we do this, and how it could be better.

**Comment by tvolkert** [comment]
Thanks @sigurdm - can you publish this to Pub too?

**Comment by sigurdm** [comment]
Yep - done

**Pull request merged**